Offline: Update/sync moar invalidation

rralian opened the pull request:
This PR addresses the issue raised here (https://github.com/Automattic/wp-calypso/pull/3696#issuecomment-195050155) and is necessary before merging #3696.

The crux of this PR is to change how we process local and server responses so that new post-lists from a local response are treated as temporary and are re-evaluated when we get the server response for that request.

This has the corner case potential to get out of whack if/when the server responses come in out of order. If that happens, we just wipe the cached list clean so the app has the ability to rebuild the list from scratch.

I am also renaming some variables to be a little more explicit, like `listKey` instead of `key` and `requestKey`, and `newPosts` and `removedPosts`, etc.

/cc @retrofox @gwwar

retrofox commented on client/lib/posts/post-list-cache-store.js:
```diff
@@ -1,73 +1,63 @@
+	// Delete the dirty cache to force a request for new data
+	if ( _cache[ listKey ] && _cache[ listKey ].dirty ) {
+		debug( 'delete cached list %o', listKey );
+		delete _cache[ listKey ];
```
can we modify a constant here?
I'm not sure but I think If `_cache_` needs changes its value you should use `let` instead of `const`.

rralian replied:
You can modify a constant. It's constant in the sense that the variable is a constant reference to the same object. It doesn't mean that the object itself cannot be changed, but just that the assignment cannot be made again. So this usage is valid. It would be invalid for example if I did something like this:
```js
const example = someObject;
example = omit( example, 'prop' );
```

retrofox replied:
You're right, the _constant_ here is the reference from the variable to the object. That reference is constant.

retrofox commented on client/lib/wp/sync-handler/index.js:
```diff
@@ -91,13 +91,14 @@ export class SyncHandler {
+				if ( localRecord && localRecord.body ) {
 					debug( 'local callback run => %o, params (%o), response (%o)', path, reqParams, localRecord );
 					// try/catch in case cached record does not match expected schema
+					localRecord.body._headers = { requestKey, responseSource: 'local'};
```
I'm pretty sure if we should use `_headers` to add local vars since it's part of the server-response.
What do you think to add another key into the body?

```es6
localRecord.body.__sync = {  requestKey, responseSource: 'local' };
```

rralian replied:
good point... I'll make that change and rebase

gwwar commented on client/lib/posts/post-list-cache-store.js:
```diff
@@ -128,4 +116,49 @@ PostsListCache.dispatchToken = Dispatcher.register( function( payload ) {
+
+	keys.forEach( function( key ) {
+		if ( cacheKey.length ) {
+			cacheKey += ':';
```
You can alternatively make CacheKey a list:
```
const cacheKey = [];
keys.forEach( function( key ) {
    cacheKey.push( `${ key }-${ options[ key ] }` );
} );
return cacheKey.join( ':' );
```

rralian replied:
for sure.

gwwar commented on client/lib/posts/post-list-cache-store.js:
```diff
@@ -1,71 +1,66 @@
-	if ( _cache[ key ] && ! isStale( _cache[ key ] ) && ! _cache[ key ].dirty ) {
-		return _cache[ key ].list;
+function get( listKey ) {
+	if ( _cache[ listKey ] && ! isStale( _cache[ listKey ] ) && ! _cache[ listKey ].dirty ) {
```
This seems long enough to move into a helper check, maybe something like
`isListKeyFresh`

rralian replied:
great point... this isn't easy to read on its own

gwwar commented on client/lib/posts/post-list-cache-store.js:
```diff
@@ -1,71 +1,66 @@
+import sitesFactory from 'lib/sites-list';
+import Dispatcher from 'dispatcher';
+
+let _cache = {};
```
Do we need private prefixes if we never export this?

rralian replied:
nope, we do not.

gwwar commented:
:+1: Verified that issues in https://github.com/Automattic/wp-calypso/pull/3696#issuecomment-195050155 are no longer present. I left some optional notes. I also noticed some pill post count jumping, but shouldn't be a blocker. Feel free to :ship: after you get a green build.

rralian commented:
Thanks @gwwar and @retrofox, I implemented your suggestions... I also fixed the post-counts jumping around. I had also noticed an issue where a Post could be shown with the wrong status due to an update from a localResponse. So now we don't update an existing object (post or post-counts) from a local response. I've squashed and am going to merge this... it's still only enabled in pre-production environments.
